bundles/org.openhab.binding.smhi/src/main/java/org/openhab/binding/smhi/internal/SmhiHandler.java

```diff
@@ -0,0 +1,350 @@
+    }
+
+    @Override
+    public void handleConfigurationUpdate(Map<String, Object> configurationParameters) {
+        if (!isModifyingCurrentConfig(configurationParameters)) {
+            return;
+        }
+
+        validateConfigurationParameters(configurationParameters);
+
+        Configuration configuration = editConfiguration();
+        for (Map.Entry<String, Object> configurationParameter : configurationParameters.entrySet()) {
+            configuration.put(configurationParameter.getKey(), configurationParameter.getValue());
+        }
+
+        if (isInitialized()) {
+            updateConfiguration(configuration);
+            config = configuration.as(SmhiConfiguration.class);
+            updateThing(editThing().withChannels(createChannels()).build());
+        } else {
+            // persist new configuration and notify Thing Manager
+            updateConfiguration(configuration);
+            ThingHandlerCallback callback = getCallback();
+            if (callback != null) {
+                callback.configurationUpdated(this.getThing());
+            } else {
+                logger.warn("Handler {} tried updating its configuration although the handler was already disposed.",
+                        this.getClass().getSimpleName());
+            }
+        }
+    }
+
+    /**
+     * Update channels with new forecast data.
+     * @param timeSeries A {@link TimeSeries} object containing forecasts.
+     */
+    private void updateChannels(TimeSeries timeSeries) {
+        // Loop through hourly forecasts and update those available
+        for (int i = 0; i < 25; i++) {
+            List<Channel> channels = thing.getChannelsOfGroup("hour_" + i);
+            if (channels.isEmpty()) {
+                continue;
+            }
+            Forecast forecast = timeSeries.getForecast(i);
+            channels.forEach(c -> {
+                String id = c.getUID().getIdWithoutGroup();
+                BigDecimal value = forecast.getParameter(id);
+                if (value != null) {
+                    // Smhi returns -9 for spp if there's no precipitation, convert to UNDEF
+                    if (id.equals(PERCENT_FROZEN) && value.intValue() == -9) {
+                        updateState(c.getUID(), UnDefType.UNDEF);
+                    } else {
+                        updateState(c.getUID(), new DecimalType(value));
+                    }
+                } else {
+                    updateState(c.getUID(), UnDefType.NULL);
+                }
+            });
+        }
+        // Loop through daily forecasts and updates those available
+        for (int i = 0; i < 10; i++) {
+            List<Channel> channels = thing.getChannelsOfGroup("day_" + i);
+            if (channels.isEmpty()) {
+                continue;
+            }
+            Forecast forecast;
+            try {
+                forecast = timeSeries.getForecast(currentDay, 24 * i + 12);
+            } catch (NoSuchElementException e) {
```

Comment:
I thought this was the most appropriate exception to throw, since it's what's being thrown by `Iterator.next()` when there are no more elements (which is basically what happens here if the time is too far off). But can change to something else.

Reply:
`NoSuchElementException` is an unchecked Exception, which should only raise if there is a bug in your code. Returning null is ok, IMHO. You could also return Optional\<Forecast>. And if there is no Forecast available, you return Optional.empty().